pkg/provisioner/resourcehandlers.go

```diff
@@ -452,35 +451,44 @@ func deleteObjectBucket(ob *v1alpha1.ObjectBucket, c versioned.Interface) error
-		return (err == nil), err
-	})
-	return
+	result, err = c.ObjectbucketV1alpha1().ObjectBucketClaims(obc.Namespace).Update(context.TODO(), obc, metav1.UpdateOptions{})
+	if err != nil {
+		// return input obc here since result is nil on error returns
+		return obc, fmt.Errorf("failed to update OBC %s/%s: %v", obc.Namespace, obc.Name, err)
+	}
+	return result, err
 }
 
 func updateObjectBucketClaimPhase(c versioned.Interface, obc *v1alpha1.ObjectBucketClaim, phase v1alpha1.ObjectBucketClaimStatusPhase, retryInterval, retryTimeout time.Duration) (result *v1alpha1.ObjectBucketClaim, err error) {
 	logD.Info("updating status:", "obc", obc.Namespace+"/"+obc.Name, "old status",
 		obc.Status.Phase, "new status", phase)
-	obc.Status.Phase = phase
+	// Do not make changes directly to the obc used as input. If the update fails, we should return
+	// the obc given as input as it was given so code that comes after can't assume obc is at the
+	// new phase.
+	updateOBC := obc.DeepCopy()
+	updateOBC.Status.Phase = phase
 
-	err = wait.PollImmediate(retryInterval, retryTimeout, func() (bool, error) {
-		result, err = c.ObjectbucketV1alpha1().ObjectBucketClaims(obc.Namespace).UpdateStatus(context.TODO(), obc, metav1.UpdateOptions{})
-		return (err == nil), err
-	})
-	return
+	result, err = c.ObjectbucketV1alpha1().ObjectBucketClaims(obc.Namespace).UpdateStatus(context.TODO(), updateOBC, metav1.UpdateOptions{})
+	if err != nil {
+		// return input obc here since result is nil on error returns
+		return obc, fmt.Errorf("failed to update OBC %s/%s phase to %q: %v", obc.Namespace, obc.Name, phase, err)
+	}
+	return result, err
 }
 
 func updateObjectBucketPhase(c versioned.Interface, ob *v1alpha1.ObjectBucket, phase v1alpha1.ObjectBucketStatusPhase, retryInterval, retryTimeout time.Duration) (result *v1alpha1.ObjectBucket, err error) {
-	logD.Info("updating status:", "ob", ob.Name, "old status", ob.Status.Phase,
-		"new status", phase)
-	ob.Status.Phase = phase
-
-	err = wait.PollImmediate(retryInterval, retryTimeout, func() (bool, error) {
-		result, err = c.ObjectbucketV1alpha1().ObjectBuckets().UpdateStatus(context.TODO(), ob, metav1.UpdateOptions{})
-		return err == nil, err
-	})
-	return
+	logD.Info("updating status:", "ob", ob.Name, "old status", ob.Status.Phase, "new status", phase)
+	// Do not make changes directly to the ob used as input. If the update fails, we should return
+	// the ob given as input as it was given so code that comes after can't assume ob is at the new
+	// phase.
+	updateOB := ob.DeepCopy()
+	updateOB.Status.Phase = phase
+
+	result, err = c.ObjectbucketV1alpha1().ObjectBuckets().UpdateStatus(context.TODO(), updateOB, metav1.UpdateOptions{})
+	if err != nil {
+		// return input ob here since result is nil on error returns
+		return ob, fmt.Errorf("failed to update OB %s phase to %q: %v", ob.Name, phase, err)
+	}
+	return result, err
```

Comment:
I took Jeff's suggestion to make the resource handler functions handle the logic for not allowing ob/obc to become nil. I think this is more clean, but there is some risk in the future if new functions don't follow the same trend. 

At the same time, I took one of Jeff's older suggestions to remove polling. I think this makes the changes I am making here a lot easier to understand conceptually.